File: server/services/ai-review.ts
/**
 * AI-extraction review workflow (US-009)
 *
 * AI-extracted drafts (from the URL extractor US-004 and the text extractor
 * US-008) are queued as `pending` contributions flagged
 * `entityData.aiGenerated` with a `entityData.perFieldConfidence` map. This
 * module turns those drafts into a **field-level review** model — a human can
 * accept / edit / reject each field before commit, low-confidence fields are
 * flagged — and, on approval, **promotes** the accepted draft into the
 * appropriate `data/source/lexicons/*.tsv` with provenance recording BOTH the AI source and
 * the human reviewer.
 *
 * Everything here is pure over an injectable `lexiconsDir` (+ a `now` clock), so
 * the whole workflow is unit-tested against `fs.mkdtempSync` temp dirs — no live
 * lexicon writes, no wall-clock.
 */

import fs from "fs";
import path from "path";
import type { Contribution, ContributionEntityType } from "./contribution-service";

// ============================================================================
// Field-level review model
// ============================================================================

/** Per-field confidence at/below this (0..1) is highlighted as low-confidence. */
export const LOW_CONFIDENCE_THRESHOLD = 0.5;

/**
 * Keys in `entityData` that are provenance/metadata, not reviewable content
 * fields. Everything else in `entityData` is projected as a review field.
 */
const METADATA_KEYS = new Set<string>([
  "source",
  "aiGenerated",
  "autoDerived",
  "provenanceKind",
  "wikidataQid",
  "sourceUrl",
  "relationships",
  "perFieldConfidence",
]);

export interface ReviewFieldView {
  field: string;
  value: unknown;
  /** 0..1 per-field confidence, or null when the extractor recorded none. */
  confidence: number | null;
  lowConfidence: boolean;
}

export interface AiDraftView {
  id: string;
  entityType: ContributionEntityType;
  action: Contribution["action"];
  status: Contribution["status"];
  aiGenerated: boolean;
  aiSource: string | null;
  /** 1..100 overall confidence (from the contribution record). */
  overallConfidence: number;
  submittedAt: string;
  reviewedAt?: string;
  reviewer?: string;
  fields: ReviewFieldView[];
  relationships: unknown[];
  promotable: boolean;
  promotion?: PromotionRecord;
}

/** True when a contribution is an AI-extracted draft (US-004 / US-008). */
export function isAiDraft(c: Contribution): boolean {
  const data = c.entityData as Record<string, unknown> | undefined;
  return !!data && data.aiGenerated === true;
}

function perFieldConfidence(c: Contribution): Record<string, number> {
  const raw = (c.entityData as Record<string, unknown>)?.perFieldConfidence;
  return raw && typeof raw === "object" ? (raw as Record<string, number>) : {};
}

/**
 * Project an AI draft into a field-level review view. Pure.
 */
export function projectDraft(c: Contribution): AiDraftView {
  const data = (c.entityData ?? {}) as Record<string, unknown>;
  const conf = perFieldConfidence(c);

  const fields: ReviewFieldView[] = Object.keys(data)
    .filter((k) => !METADATA_KEYS.has(k))
    .map((field) => {
      const confidence = typeof conf[field] === "number" ? conf[field] : null;
      return {
        field,
        value: data[field],
        confidence,
        lowConfidence: confidence !== null && confidence < LOW_CONFIDENCE_THRESHOLD,
      };
    });

  return {
    id: c.id,
    entityType: c.entityType,
    action: c.action,
    status: c.status,
    aiGenerated: isAiDraft(c),
    aiSource: typeof data.source === "string" ? data.source : null,
    overallConfidence: c.confidence,
    submittedAt: c.submittedAt,
    reviewedAt: c.reviewedAt,
    reviewer: c.reviewer,
    fields,
    relationships: Array.isArray(data.relationships) ? (data.relationships as unknown[]) : [],
    promotable: c.entityType in PROMOTION_TARGETS,
    promotion: c.promotion,
  };
}

// ============================================================================
// Field decisions → accepted entityData
// ============================================================================

export type FieldDecision =
  | { decision: "accept" }
  | { decision: "edit"; value: unknown }
  | { decision: "reject" };

export interface FieldReviewRecord {
  decision: "accept" | "edit" | "reject";
  value?: unknown;
}

export interface AppliedReview {
  /** Content fields after applying the per-field decisions (no metadata). */
  acceptedData: Record<string, unknown>;
  fieldReviews: Record<string, FieldReviewRecord>;
  rejectedFields: string[];
}

/** Minimal required content fields for the promotable entity types. */
const REQUIRED_CONTENT: Partial<Record<ContributionEntityType, string[]>> = {
  civilization: ["name"],
  language: ["name"],
  "archaeological-site": ["name", "coordinates"],
  "trade-good": ["name"],
  "historical-figure": ["name"],
};

export class AiReviewError extends Error {}

/**
 * Apply per-field accept/edit/reject decisions to an AI draft. Fields with no
 * explicit decision default to `accept`. Returns the accepted content fields
 * (metadata stripped) plus the recorded decisions. Pure.
 *
 * Throws `AiReviewError` if a decision names a field that isn't in the draft.
 */
export function applyFieldReviews(
  c: Contribution,
  decisions: Record<string, FieldDecision> = {},
): AppliedReview {
  const view = projectDraft(c);
  const known = new Set(view.fields.map((f) => f.field));

  for (const field of Object.keys(decisions)) {
    if (!known.has(field)) {
      throw new AiReviewError(`Unknown field '${field}' for contribution ${c.id}`);
    }
  }

  const acceptedData: Record<string, unknown> = {};
  const fieldReviews: Record<string, FieldReviewRecord> = {};
  const rejectedFields: string[] = [];

  for (const f of view.fields) {
    const decision = decisions[f.field] ?? { decision: "accept" as const };
    if (decision.decision === "reject") {
      fieldReviews[f.field] = { decision: "reject" };
      rejectedFields.push(f.field);
      continue;
    }
    if (decision.decision === "edit") {
      acceptedData[f.field] = decision.value;
      fieldReviews[f.field] = { decision: "edit", value: decision.value };
      continue;
    }
    acceptedData[f.field] = f.value;
    fieldReviews[f.field] = { decision: "accept" };
  }

  return { acceptedData, fieldReviews, rejectedFields };
}

/**
 * Validate that the accepted content satisfies the target entity type's
 * required fields (e.g. a reviewer can't reject `name`). Returns error strings.
 */
export function validateAcceptedDraft(
  entityType: ContributionEntityType,
  acceptedData: Record<string, unknown>,
): string[] {
  const errors: string[] = [];
  const required = REQUIRED_CONTENT[entityType] ?? ["name"];
  for (const field of required) {
    if (!(field in acceptedData) || acceptedData[field] === undefined || acceptedData[field] === "") {
      errors.push(`Required field '${field}' was rejected or is empty`);
    }
  }
  return errors;
}

// ============================================================================
// Promotion to data/source/lexicons/*.tsv
// ============================================================================

export interface PromotionRecord {
  file: string;
  targetId: string;
  aiSource: string;
  reviewer: string;
  reviewedAt: string;
}

interface PromotionTarget {
  file: string;
  header: string[];
  /** Map accepted content fields → header-column cells (strings). */
  map: (data: Record<string, unknown>, overallConfidence: number) => Record<string, string>;
}

function coordCell(value: unknown): string {
  if (value && typeof value === "object") {
    const c = value as { lat?: unknown; lng?: unknown };
    if (typeof c.lat === "number" && typeof c.lng === "number") {
      return JSON.stringify({ lat: c.lat, lng: c.lng });
    }
  }
  return "";
}

function timePeriodRange(data: Record<string, unknown>): string {
  const start = data.timePeriodStart;
  const end = data.timePeriodEnd;
  if (start === undefined && end === undefined) return "";
  return `${start ?? ""} to ${end ?? ""}`.trim();
}

function str(v: unknown): string {
  if (v === undefined || v === null) return "";
  return String(v);
}

const PROMOTION_TARGETS: Partial<Record<ContributionEntityType, PromotionTarget>> = {
  civilization: {
    file: "civilizations.tsv",
    header: [
      "id", "name", "native_name", "time_period_start", "time_period_end",
      "time_period_label", "associated_language_ids", "writing_systems",
      "political_structure", "capital", "population", "haplogroup_ids",
      "cuisine_id", "sources", "description",
    ],
    map: (data) => ({
      name: str(data.name),
      time_period_start: str(data.timePeriodStart),
      time_period_end: str(data.timePeriodEnd),
      description: str(data.description),
    }),
  },
  language: {
    file: "languages.tsv",
    header: [
      "id", "name", "native_name", "iso639_1", "iso639_2", "family_id",
      "parent_language_id", "region", "countries", "native_speakers",
      "total_speakers", "status", "time_origin", "time_end", "classification",
      "writing_system", "is_historical_variant", "is_dialect",
      "chronological_order", "historical_context", "latitude", "longitude",
    ],
    map: (data) => {
      const cell: Record<string, string> = {
        name: str(data.name),
        historical_context: str(data.description),
        time_origin: str(data.timePeriodStart),
        time_end: str(data.timePeriodEnd),
      };
      const coord = data.coordinates as { lat?: unknown; lng?: unknown } | undefined;
      if (coord && typeof coord.lat === "number" && typeof coord.lng === "number") {
        cell.latitude = String(coord.lat);
        cell.longitude = String(coord.lng);
      }
      return cell;
    },
  },
  "archaeological-site": {
    file: "archaeological-sites.tsv",
    header: [
      "id", "name", "coordinates", "site_type", "time_period_start",
      "time_period_end", "time_period_label", "associated_language_ids",
      "associated_culture_ids", "associated_civilization_ids",
      "excavation_status", "findings", "importance", "confidence", "sources",
      "description",
    ],
    map: (data, overallConfidence) => ({
      name: str(data.name),
      coordinates: coordCell(data.coordinates),
      time_period_start: str(data.timePeriodStart),
      time_period_end: str(data.timePeriodEnd),
      confidence: String(overallConfidence),
      description: str(data.description),
    }),
  },
  "trade-good": {
    file: "trade-goods.tsv",
    header: [
      "id", "name", "category", "origin_region", "origin_coordinates",
      "trade_routes", "time_period", "economic_significance",
      "associated_languages",
    ],
    map: (data) => ({
      name: str(data.name),
      origin_coordinates: coordCell(data.coordinates),
      time_period: timePeriodRange(data),
      economic_significance: str(data.description),
    }),
  },
};

const PROVENANCE_LEDGER = "contribution-provenance.tsv";
const PROVENANCE_HEADER = [
  "contribution_id", "entity_type", "target_file", "target_id",
  "ai_source", "reviewer", "reviewed_at", "confidence",
];

function slugify(value: string): string {
  return value
    .toLowerCase()
    .normalize("NFKD")
    .replace(/[^\w\s-]/g, "")
    .trim()
    .replace(/[\s_]+/g, "-")
    .replace(/-+/g, "-") || "entity";
}

function readExistingIds(filePath: string): Set<string> {
  const ids = new Set<string>();
  if (!fs.existsSync(filePath)) return ids;
  const lines = fs.readFileSync(filePath, "utf-8").split(/\r?\n/);
  for (let i = 1; i < lines.length; i++) {
    if (!lines[i]) continue;
    ids.add(lines[i].split("\t")[0]);
  }
  return ids;
}

function uniqueId(base: string, existing: Set<string>): string {
  if (!existing.has(base)) return base;
  let n = 2;
  while (existing.has(`${base}-${n}`)) n++;
  return `${base}-${n}`;
}

/**
 * Append a row (aligned to `header`) to a TSV, creating the file with its header
 * when absent. Preserves a trailing newline.
 */
function appendRow(filePath: string, header: string[], cells: Record<string, string>): void {
  const row = header.map((col) => (cells[col] ?? "").replace(/[\t\n\r]/g, " ")).join("\t");
  if (!fs.existsSync(filePath)) {
    fs.writeFileSync(filePath, `${header.join("\t")}\n${row}\n`, "utf-8");
    return;
  }
  const existing = fs.readFileSync(filePath, "utf-8");
  const sep = existing.length > 0 && !existing.endsWith("\n") ? "\n" : "";
  fs.appendFileSync(filePath, `${sep}${row}\n`, "utf-8");
}

/**
 * Promote an accepted AI draft into the appropriate `data/source/lexicons/*.tsv` and record
 * provenance (AI source + human reviewer) in a companion ledger. Returns the
 * promotion record. Throws `AiReviewError` for a non-promotable entity type or
 * when a required field is missing.
 */
export function promoteContribution(params: {
  contributionId: string;
  entityType: ContributionEntityType;
  acceptedData: Record<string, unknown>;
  reviewer: string;
  aiSource: string;
  overallConfidence: number;
  lexiconsDir: string;
  now: string;
}): PromotionRecord {
  const target = PROMOTION_TARGETS[params.entityType];
  if (!target) {
    throw new AiReviewError(`No TSV promotion target for entity type '${params.entityType}'`);
  }

  const errors = validateAcceptedDraft(params.entityType, params.acceptedData);
  if (errors.length > 0) {
    throw new AiReviewError(errors.join("; "));
  }

  const filePath = path.join(params.lexiconsDir, target.file);
  const existingIds = readExistingIds(filePath);
  const name = str(params.acceptedData.name);
  const targetId = uniqueId(slugify(name), existingIds);

  const cells = target.map(params.acceptedData, params.overallConfidence);
  cells.id = targetId;
  // Provenance inline where the file has a `sources` column (string array shape).
  if (target.header.includes("sources")) {
    cells.sources = JSON.stringify([
      `AI-extracted via ${params.aiSource}; reviewed by ${params.reviewer}`,
    ]);
  }

  appendRow(filePath, target.header, cells);

  // Structured provenance ledger — uniform regardless of the target's columns.
  appendRow(path.join(params.lexiconsDir, PROVENANCE_LEDGER), PROVENANCE_HEADER, {
    contribution_id: params.contributionId,
    entity_type: params.entityType,
    target_file: target.file,
    target_id: targetId,
    ai_source: params.aiSource,
    reviewer: params.reviewer,
    reviewed_at: params.now,
    confidence: String(params.overallConfidence),
  });

  return {
    file: target.file,
    targetId,
    aiSource: params.aiSource,
    reviewer: params.reviewer,
    reviewedAt: params.now,
  };
}

export function isPromotable(entityType: ContributionEntityType): boolean {
  return entityType in PROMOTION_TARGETS;
}

File: services/api/src/pinakes/contributions/ai_review.py
"""Field-level review of AI drafts, and their promotion into the lexicon TSVs.

The port of `server/services/ai-review.ts`. AI-extracted drafts (the URL and text
extractors) sit in the queue flagged ``entityData.aiGenerated`` with a per-field
confidence map; this module turns one into a **field-level review model** — a
human accepts, edits, or rejects each field — and, on approval, appends the
accepted content to the matching ``data/source/lexicons/*.tsv`` with provenance
naming *both* the AI source and the human reviewer.

Everything is pure over an explicit ``lexicons_dir`` and ``now``, so the whole
workflow is exercised against a temp directory. That injectability is not a
nicety here: a test that promoted into the live corpus would leave a fixture row
visible to every other reader of it (`server/CLAUDE.md`).

One deliberate divergence from the TypeScript original: a required field whose
accepted value is ``null`` fails validation here, where JS only rejected
``undefined`` and ``""``. A JSON ``null`` and an absent value are the same value
in Python, and promoting a null `name` would write an empty id into the corpus —
so the stricter reading is the one that keeps the corpus honest.
"""

from __future__ import annotations

import json
import re
import unicodedata
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from pinakes.contributions.store import Contribution

#: Per-field confidence at or below this (0..1) is flagged for the reviewer.
LOW_CONFIDENCE_THRESHOLD = 0.5

#: `entityData` keys that are provenance/metadata rather than reviewable
#: content. Everything else in `entityData` is projected as a review field.
METADATA_KEYS = frozenset(
    {
        "source",
        "aiGenerated",
        "autoDerived",
        "provenanceKind",
        "wikidataQid",
        "sourceUrl",
        "relationships",
        "perFieldConfidence",
    }
)

#: Minimal required content fields per promotable entity type.
REQUIRED_CONTENT: dict[str, tuple[str, ...]] = {
    "civilization": ("name",),
    "language": ("name",),
    "archaeological-site": ("name", "coordinates"),
    "trade-good": ("name",),
    "historical-figure": ("name",),
}

#: The companion ledger every promotion is recorded in, whatever the target's
#: own columns look like.
PROVENANCE_LEDGER = "contribution-provenance.tsv"
PROVENANCE_HEADER = (
    "contribution_id",
    "entity_type",
    "target_file",
    "target_id",
    "ai_source",
    "reviewer",
    "reviewed_at",
    "confidence",
)


class AiReviewError(Exception):
    """A review that cannot be applied — a 400, never a 500."""


@dataclass(frozen=True, slots=True)
class AppliedReview:
    """The outcome of applying per-field decisions to a draft."""

    accepted_data: dict[str, Any]
    field_reviews: dict[str, dict[str, Any]]
    rejected_fields: list[str]


# ── Projection ───────────────────────────────────────────────────────────────


def is_ai_draft(contribution: Contribution) -> bool:
    """True when a contribution is an AI-extracted draft."""
    entity_data = contribution.get("entityData")
    return isinstance(entity_data, dict) and entity_data.get("aiGenerated") is True


def _per_field_confidence(contribution: Contribution) -> dict[str, Any]:
    entity_data = contribution.get("entityData")
    if not isinstance(entity_data, dict):
        return {}
    raw = entity_data.get("perFieldConfidence")
    return raw if isinstance(raw, dict) else {}


def project_draft(contribution: Contribution) -> dict[str, Any]:
    """Project a draft into the field-level review view the client renders.

    Insertion order of `entityData` is the field order, which is the order the
    extractor wrote — same as `Object.keys` gives the TypeScript projection.
    """
    entity_data = contribution.get("entityData")
    data: dict[str, Any] = entity_data if isinstance(entity_data, dict) else {}
    confidences = _per_field_confidence(contribution)

    fields: list[dict[str, Any]] = []
    for field in data:
        if field in METADATA_KEYS:
            continue
        raw = confidences.get(field)
        confidence: float | None = None
        if isinstance(raw, (int, float)) and not isinstance(raw, bool):
            confidence = float(raw)
        fields.append(
            {
                "field": field,
                "value": data[field],
                "confidence": confidence,
                "lowConfidence": confidence is not None
                and confidence < LOW_CONFIDENCE_THRESHOLD,
            }
        )

    source = data.get("source")
    relationships = data.get("relationships")
    view: dict[str, Any] = {
        "id": contribution.get("id"),
        "entityType": contribution.get("entityType"),
        "action": contribution.get("action"),
        "status": contribution.get("status"),
        "aiGenerated": is_ai_draft(contribution),
        "aiSource": source if isinstance(source, str) else None,
        "overallConfidence": contribution.get("confidence"),
        "submittedAt": contribution.get("submittedAt"),
        "reviewedAt": contribution.get("reviewedAt"),
        "reviewer": contribution.get("reviewer"),
        "fields": fields,
        "relationships": relationships if isinstance(relationships, list) else [],
        "promotable": is_promotable(str(contribution.get("entityType"))),
        "promotion": contribution.get("promotion"),
    }
    # `reviewedAt`, `reviewer` and `promotion` are optional on the TypeScript
    # view and absent — not null — until a reviewer has acted.
    for optional in ("reviewedAt", "reviewer", "promotion"):
        if view[optional] is None:
            del view[optional]
    return view


# ── Field decisions ──────────────────────────────────────────────────────────


def apply_field_reviews(
    contribution: Contribution, decisions: dict[str, Any] | None = None
) -> AppliedReview:
    """Apply accept/edit/reject decisions; an undecided field is accepted.

    Raises :class:`AiReviewError` if a decision names a field the draft does not
    have — a reviewer editing a field that is not there is a client bug worth
    reporting, not a silent no-op.
    """
    decisions = decisions or {}
    view = project_draft(contribution)
    known = {field["field"] for field in view["fields"]}

    for field in decisions:
        if field not in known:
            raise AiReviewError(
                f"Unknown field '{field}' for contribution {contribution.get('id')}"
            )

    accepted: dict[str, Any] = {}
    reviews: dict[str, dict[str, Any]] = {}
    rejected: list[str] = []

    for entry in view["fields"]:
        field = entry["field"]
        decision = decisions.get(field) or {"decision": "accept"}
        kind = decision.get("decision") if isinstance(decision, dict) else "accept"
        if kind == "reject":
            reviews[field] = {"decision": "reject"}
            rejected.append(field)
            continue
        if kind == "edit":
            value = decision.get("value") if isinstance(decision, dict) else None
            accepted[field] = value
            reviews[field] = {"decision": "edit", "value": value}
            continue
        accepted[field] = entry["value"]
        reviews[field] = {"decision": "accept"}

    return AppliedReview(
        accepted_data=accepted, field_reviews=reviews, rejected_fields=rejected
    )


def validate_accepted_draft(
    entity_type: str, accepted_data: dict[str, Any]
) -> list[str]:
    """Required fields the reviewer rejected or emptied. Empty list ⇒ promotable."""
    errors: list[str] = []
    for field in REQUIRED_CONTENT.get(entity_type, ("name",)):
        value = accepted_data.get(field)
        if field not in accepted_data or value is None or value == "":
            errors.append(f"Required field '{field}' was rejected or is empty")
    return errors


# ── Promotion targets ────────────────────────────────────────────────────────


def _str(value: Any) -> str:
    """``String(v)`` for a cell, with null/undefined rendering empty.

    A whole-number float renders without its ``.0``: JSON has one number type and
    JavaScript prints `3000.0` as `3000`, so a year read back out of a draft must
    not gain a decimal point on its way into the corpus.
    """
    if value is None:
        return ""
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value)


def _coord_cell(value: Any) -> str:
    """The corpus's coordinate cell — `{"lat":..,"lng":..}` or empty."""
    if isinstance(value, dict):
        lat = value.get("lat")
        lng = value.get("lng")
        if _is_number(lat) and _is_number(lng):
            return json.dumps({"lat": lat, "lng": lng}, separators=(",", ":"))
    return ""


def _is_number(value: Any) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def _time_period_range(data: dict[str, Any]) -> str:
    start = data.get("timePeriodStart")
    end = data.get("timePeriodEnd")
    if "timePeriodStart" not in data and "timePeriodEnd" not in data:
        return ""
    return f"{_str(start)} to {_str(end)}".strip()


def _civilization_cells(data: dict[str, Any], _confidence: Any) -> dict[str, str]:
    return {
        "name": _str(data.get("name")),
        "time_period_start": _str(data.get("timePeriodStart")),
        "time_period_end": _str(data.get("timePeriodEnd")),
        "description": _str(data.get("description")),
    }


def _language_cells(data: dict[str, Any], _confidence: Any) -> dict[str, str]:
    cells = {
        "name": _str(data.get("name")),
        "historical_context": _str(data.get("description")),
        "time_origin": _str(data.get("timePeriodStart")),
        "time_end": _str(data.get("timePeriodEnd")),
    }
    coord = data.get("coordinates")
    if isinstance(coord, dict) and _is_number(coord.get("lat")) and _is_number(
        coord.get("lng")
    ):
        cells["latitude"] = _str(coord["lat"])
        cells["longitude"] = _str(coord["lng"])
    return cells


def _site_cells(data: dict[str, Any], confidence: Any) -> dict[str, str]:
    return {
        "name": _str(data.get("name")),
        "coordinates": _coord_cell(data.get("coordinates")),
        "time_period_start": _str(data.get("timePeriodStart")),
        "time_period_end": _str(data.get("timePeriodEnd")),
        "confidence": _str(confidence),
        "description": _str(data.get("description")),
    }


def _trade_good_cells(data: dict[str, Any], _confidence: Any) -> dict[str, str]:
    return {
        "name": _str(data.get("name")),
        "origin_coordinates": _coord_cell(data.get("coordinates")),
        "time_period": _time_period_range(data),
        "economic_significance": _str(data.get("description")),
    }


@dataclass(frozen=True, slots=True)
class PromotionTarget:
    """Where one entity type lands, and how its accepted fields map to columns."""

    file: str
    header: tuple[str, ...]
    cells: Callable[[dict[str, Any], Any], dict[str, str]]


#: The promotable types, and only those. `historical-figure` has no TSV of its
#: own, so it is *reviewable* but not promotable — approving one is a 400, not a
#: silent write into the wrong file.
PROMOTION_TARGETS: dict[str, PromotionTarget] = {
    "civilization": PromotionTarget(
        file="civilizations.tsv",
        header=(
            "id", "name", "native_name", "time_period_start", "time_period_end",
            "time_period_label", "associated_language_ids", "writing_systems",
            "political_structure", "capital", "population", "haplogroup_ids",
            "cuisine_id", "sources", "description",
        ),
        cells=_civilization_cells,
    ),
    "language": PromotionTarget(
        file="languages.tsv",
        header=(
            "id", "name", "native_name", "iso639_1", "iso639_2", "family_id",
            "parent_language_id", "region", "countries", "native_speakers",
            "total_speakers", "status", "time_origin", "time_end", "classification",
            "writing_system", "is_historical_variant", "is_dialect",
            "chronological_order", "historical_context", "latitude", "longitude",
        ),
        cells=_language_cells,
    ),
    "archaeological-site": PromotionTarget(
        file="archaeological-sites.tsv",
        header=(
            "id", "name", "coordinates", "site_type", "time_period_start",
            "time_period_end", "time_period_label", "associated_language_ids",
            "associated_culture_ids", "associated_civilization_ids",
            "excavation_status", "findings", "importance", "confidence", "sources",
            "description",
        ),
        cells=_site_cells,
    ),
    "trade-good": PromotionTarget(
        file="trade-goods.tsv",
        header=(
            "id", "name", "category", "origin_region", "origin_coordinates",
            "trade_routes", "time_period", "economic_significance",
            "associated_languages",
        ),
        cells=_trade_good_cells,
    ),
}


def is_promotable(entity_type: str) -> bool:
    """Whether an approved draft of this type has somewhere to be written."""
    return entity_type in PROMOTION_TARGETS


# ── Writing the row ──────────────────────────────────────────────────────────


def slugify(value: str) -> str:
    """The corpus id form: lowercase, ASCII-word characters, hyphen-joined."""
    text = unicodedata.normalize("NFKD", value.lower())
    text = re.sub(r"[^\w\s-]", "", text, flags=re.ASCII)
    text = re.sub(r"[\s_]+", "-", text.strip())
    return re.sub(r"-+", "-", text) or "entity"


def _existing_ids(path: Path) -> set[str]:
    """The first column of every data row — a TSV's id set."""
    if not path.is_file():
        return set()
    lines = path.read_text(encoding="utf-8").splitlines()
    return {line.split("\t")[0] for line in lines[1:] if line}


def _unique_id(base: str, existing: set[str]) -> str:
    if base not in existing:
        return base
    index = 2
    while f"{base}-{index}" in existing:
        index += 1
    return f"{base}-{index}"


def append_row(path: Path, header: tuple[str, ...], cells: dict[str, str]) -> None:
    """Append one row aligned to *header*, creating the file with it when absent.

    Tabs and newlines inside a cell become spaces — a TSV has no quoting, so a
    stray one would silently split the row into two.
    """
    row = "\t".join(
        re.sub(r"[\t\n\r]", " ", cells.get(column, "")) for column in header
    )
    if not path.is_file():
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text("\t".join(header) + "\n" + row + "\n", encoding="utf-8")
        return
    existing = path.read_text(encoding="utf-8")
    separator = "\n" if existing and not existing.endswith("\n") else ""
    with path.open("a", encoding="utf-8") as handle:
        handle.write(separator + row + "\n")


def promote_contribution(
    *,
    contribution_id: str,
    entity_type: str,
    accepted_data: dict[str, Any],
    reviewer: str,
    ai_source: str,
    overall_confidence: Any,
    lexicons_dir: Path,
    now: str,
) -> dict[str, Any]:
    """Append an accepted draft to its lexicon and record its provenance.

    Provenance is written twice on purpose: inline in the target's own ``sources``
    column when it has one (so a reader of that file sees it), and as a row in
    the companion ledger (so the record is uniform across targets whose columns
    differ). Raises :class:`AiReviewError` for a non-promotable type or a missing
    required field.
    """
    target = PROMOTION_TARGETS.get(entity_type)
    if target is None:
        raise AiReviewError(
            f"No TSV promotion target for entity type '{entity_type}'"
        )

    errors = validate_accepted_draft(entity_type, accepted_data)
    if errors:
        raise AiReviewError("; ".join(errors))

    path = Path(lexicons_dir) / target.file
    target_id = _unique_id(
        slugify(_str(accepted_data.get("name"))), _existing_ids(path)
    )

    cells: dict[str, str] = target.cells(accepted_data, overall_confidence)
    cells["id"] = target_id
    if "sources" in target.header:
        cells["sources"] = json.dumps(
            [f"AI-extracted via {ai_source}; reviewed by {reviewer}"],
            separators=(",", ":"),
        )

    append_row(path, target.header, cells)
    append_row(
        Path(lexicons_dir) / PROVENANCE_LEDGER,
        PROVENANCE_HEADER,
        {
            "contribution_id": contribution_id,
            "entity_type": entity_type,
            "target_file": target.file,
            "target_id": target_id,
            "ai_source": ai_source,
            "reviewer": reviewer,
            "reviewed_at": now,
            "confidence": _str(overall_confidence),
        },
    )

    return {
        "file": target.file,
        "targetId": target_id,
        "aiSource": ai_source,
        "reviewer": reviewer,
        "reviewedAt": now,
    }
